### ISSUE TITLE
EVENT_RECEIVED message

### ISSUE THREAD

**Issue by oberstet** [wamp-proto/wamp-proto#280]
WAMP has *acknowledged event publication*.

It currently does not have *acknowledged event delivery*.

Historically, this was by design:

* acknowledged event delivery requires the router to maintain _per subscriber_ queues, which comes with a high price
* most situations don't need it
* and WAMP has event history, which allows clients not only to catch up, but is in fact more powerful

However, in some situations, acknowledged event delivery is nevertheless useful.

It is also required for routers that implement a MQTT-WAMP bridge and want to provide MQTT QoS 2 on the delivery (subscriber) leg of things.

This issue first defines a new WAMP message (luckily, we have extensibility built into WAMP):

```
EVENT_RECEIVED message:

Format: [EVENT_RECEIVED, EVENT.Publication|id]
```

Currently, Autobahn/Crossbar.io implement that using implementation-specific message code 337.

Should it be ratified, the message code should be 37.

**Comment by oberstet** [comment]
Crossbar.io/Autobahn is also currently using a custom attribute `x_acknowledged_delivery`

**Comment by gammazero** [comment]
If even delivery notification is needed, then this can already be build on top of the existing pub/sub, and seems like something that would be handled at the application level, not dictated by the WAMP spec.

I would expect there would need to be some coordination between wamp clients in the application, using already available wamp functionality:

Whoever is interested in receiving events sets up a subscription to the appropriate topic(s), such as `received.event` or `received.red.alert`, `received.yellow.alert`, etc.  Clients that want to publish that they received an EVENT can do so by publishing another EVENT to one of this topics.  That even can contain whatever details need to be communicated in the acknowledgement (received timestamp, what it did with the message, etc.).

This seems like something that WAMP should not try standardize, when it already provides the mechanisms to facilitate this type of communication.

**Comment by oberstet** [comment]
One can always rebuild everything on something, eg implement RPC with PubSub using one-time topics (for call returns). It just sucks practically.

IMO an acknowledgement for event reception is something that should be implemented at the WAMP level, as it is part of the message conversation for a WAMP action (eg a publication).

I would go even further: we might not only add `EVENT_RECEIVED`, but also `PUBLICATION_DELIVERED` to signal back to the original publisher.

**Comment by gammazero** [comment]
I agree with your point.  I would like to offer another discussion in the interest of keeping the WAMP protocol as orthogonal as possible:

The EVENT_RECEIVED use case - to ensure delivery where transport may not be able to - is a fairly limited case (like MQTT bridge).  I assume that it would be the subscribers' responsibilities to request this feature when they subscribe, since only the subscribers would know if they needed it.

So, given that this is the subscribers' responsibility, those subscribers could also periodically check event history to see if they had missed anything.

The downside of using event history like this is that the special clients would be polling (generating traffic unnecessarily).  The advantage of using event history is that the burden of ensuring event delivery is pushed out to the clients that care, instead of making the router keep track of the event delivery for each of these clients.  _Using history also makes event delivery reliable across router restarts (assuming history is persisted), since sessions do not persist_.

Given the advantages, consider making event history the way to solve this reliable message delivery use case.

**Comment by gammazero** [comment]
I think `PUBLICATION_DELIVERED` is problematic with regard to WAMP.  Subscribers come and go, and a publisher should not know who the subscribers are - this is one of the nice things about WAMP versus other protocols.  In the time window between a message being published and the resulting `PUBLICATION_DELIVERED`, unless implemented as a atomic action, some subscribers may have left and others may have joined.  What is the value of `PUBLICATION_DELIVERED`?  The publisher would get a list of sessions that received the event, some of which may no longer exist.  What about the ones that did not?  Did they join just after publication or not receive the message? Did they reconnect after being disconnected (history solves this).

If `EVENT_RECEIVED` is implemented, then consider instead of having a `PUBLICATION_DELIVERED` message, have an ERROR to indicate publication to a particular session failed (`wamp.error.publication_failed`).  That seems like a more actionable message.

Still suggesting not supporting this in favor of using event history only.

**Comment by konsultaner** [comment]
@oberstet Wouln't that end up in having something equal to a RPC request with an empty `RESULT`?

**Comment by oberstet** [comment]
@gammazero this is funny=) because your argumentation vs EVENT_RECEIVED, in favor of event history because it works for the practical case is where we come from, and which I held up for a long time in quite many discussions. Because of the reasons above. Then routers then need message queues per session. And so on. It easily transforms into a can of worms;)

Rgd the pollling: actually, the event history API isn't complete, we need windowed queries on history, and then if the client is able to persist locally, it can even survive crashes, and exactly continue from where it left.

Also: either there is a proposal somewhere, or we already have it in Crossbar.io (not sure right now): an option to SUBSCRIBE to be delivered the history (only n back based today) as normal EVENTs - but only to the subscribing clients. This is to avoid races.

---

Now, all above being said, I really only see the MQTT QoS2 strict semantical mapping as a use case, and it is indeed the question if it's worth all those hassles just to have that feature in an _implementation_ of a router (the MQTT bridge thing in Crossbar.io). So yes, in this case I would say: Crossbar.io/Autobahn should keep the messages under custom message codes.

I also agree with the issues PUBLICATION_DELIVERED as "just a signal" has. Not much value in light of this. However, if the receiving side (the subscriber) can signal back _application payload_, this might be different. As in: the publication is actually a poll. Mmh.

**Comment by gammazero** [comment]
To follow up on this, now that I have had a couple of years think about it, I think the problems introduce by `PUBLICATION_DELIVERED` and `EVENT_RECEIVED`, are not worth it. So, I downvote this feature.
